app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php

```diff
@@ -442,6 +444,8 @@ private function retrieveCategoriesTree(int $storeId, array $shownCategoriesIds)
 
         $collection->addAttributeToFilter('entity_id', ['in' => array_keys($shownCategoriesIds)])
             ->addAttributeToSelect(['name', 'is_active', 'parent_id'])
+            ->addAttributeToSort('level', 'asc')
```

Comment:
'level'  and 'position' has the constant in \Magento\Catalog\Api\Data\CatalogInterface
    const KEY_POSITION = 'position';
    const KEY_LEVEL = 'level';

Reply:
thank @edenduong , I pushed a new commitI